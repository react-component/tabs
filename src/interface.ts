import { TabPaneProps } from './TabPanelList/TabPane';

export type TabSizeMap = Map<
  React.Key,
  { width: number; height: number; left: number; top: number }
>;

export interface TabOffset {
  width: number;
  height: number;
  left: number;
  right: number;
  top: number;
}
export type TabOffsetMap = Map<React.Key, TabOffset>;

export type TabPosition = 'left' | 'right' | 'top' | 'bottom';

export interface Tab extends TabPaneProps {
  key: string;
  node: React.ReactElement;
}

export type RenderTabBar = (props: any, DefaultTabBar: React.ComponentType) => React.ReactElement;

export interface TabsLocale {
  dropdownAriaLabel?: string;
  removeAriaLabel?: string;
  addAriaLabel?: string;
}

export interface EditableConfig {
  onEdit: (
    type: 'add' | 'remove',
    info: { key?: string; event: React.MouseEvent | React.KeyboardEvent },
  ) => void;
  showAdd?: boolean;
  removeIcon?: React.ReactNode;
  addIcon?: React.ReactNode;
}

export interface AnimatedConfig {
  inkBar?: boolean;
  tabPane?: boolean;
}

export type OnTabScroll = (info: { direction: 'left' | 'right' | 'top' | 'bottom' }) => void;

export type TabBarExtraPosition = 'left' | 'right';

export type TabBarExtraMap = Partial<Record<TabBarExtraPosition, React.ReactNode>>;

export type TabBarExtraContent = React.ReactNode | TabBarExtraMap;
