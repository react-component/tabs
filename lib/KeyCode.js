"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = {
  /**
   * LEFT
   */
  LEFT: 37, // also NUM_WEST
  /**
   * UP
   */
  UP: 38, // also NUM_NORTH
  /**
   * RIGHT
   */
  RIGHT: 39, // also NUM_EAST
  /**
   * DOWN
   */
  DOWN: 40 // also NUM_SOUTH
};
module.exports = exports['default'];