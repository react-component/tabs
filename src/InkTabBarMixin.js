import { setTransform, isTransformSupported } from './utils';
import React from 'react';
import classnames from 'classnames';

export function getScroll(w, top) {
  let ret = w[`page${top ? 'Y' : 'X'}Offset`];
  const method = `scroll${top ? 'Top' : 'Left'}`;
  if (typeof ret !== 'number') {
    const d = w.document;
    // ie6,7,8 standard mode
    ret = d.documentElement[method];
    if (typeof ret !== 'number') {
      // quirks mode
      ret = d.body[method];
    }
  }
  return ret;
}

function offset(elem) {
  let box;
  let x;
  let y;
  const doc = elem.ownerDocument;
  const body = doc.body;
  const docElem = doc && doc.documentElement;
  box = elem.getBoundingClientRect();
  x = box.left;
  y = box.top;
  x -= docElem.clientLeft || body.clientLeft || 0;
  y -= docElem.clientTop || body.clientTop || 0;
  const w = doc.defaultView || doc.parentWindow;
  x += getScroll(w);
  y += getScroll(w, true);
  return {
    left: x, top: y,
  };
}

function componentDidUpdate(component, init) {
  const refs = component.refs;
  const { styles } = component.props;
  const wrapNode = refs.nav || refs.root;
  const containerOffset = offset(wrapNode);
  const inkBarNode = refs.inkBar;
  const activeTab = component.activeTab;
  const inkBarNodeStyle = inkBarNode.style;
  const tabBarPosition = component.props.tabBarPosition;
  if (init) {
    // prevent mount animation
    inkBarNodeStyle.display = 'none';
  }
  if (activeTab) {
    const tabNode = activeTab;
    const tabOffset = offset(tabNode);
    const transformSupported = isTransformSupported(inkBarNodeStyle);
    if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
      let left = tabOffset.left - containerOffset.left;
      let width = tabNode.offsetWidth;
      if (styles.inkBar && styles.inkBar.width !== undefined) {
        width = parseFloat(styles.inkBar.width, 10);
        if (width) {
          left = left + (tabNode.offsetWidth - width) / 2;
        }
      }
      // use 3d gpu to optimize render
      if (transformSupported) {
        setTransform(inkBarNodeStyle, `translate3d(${left}px,0,0)`);
        inkBarNodeStyle.width = `${width}px`;
        inkBarNodeStyle.height = '';
      } else {
        inkBarNodeStyle.left = `${left}px`;
        inkBarNodeStyle.top = '';
        inkBarNodeStyle.bottom = '';
        inkBarNodeStyle.right = `${wrapNode.offsetWidth - left - width}px`;
      }
    } else {
      let top = tabOffset.top - containerOffset.top;
      let height = tabNode.offsetHeight;
      if (styles.inkBar && styles.inkBar.height !== undefined) {
        height = parseFloat(styles.inkBar.height, 10);
        if (height) {
          top = top + (tabNode.offsetHeight - height) / 2;
        }
      }
      if (transformSupported) {
        setTransform(inkBarNodeStyle, `translate3d(0,${top}px,0)`);
        inkBarNodeStyle.height = `${height}px`;
        inkBarNodeStyle.width = '';
      } else {
        inkBarNodeStyle.left = '';
        inkBarNodeStyle.right = '';
        inkBarNodeStyle.top = `${top}px`;
        inkBarNodeStyle.bottom = `${wrapNode.offsetHeight - top - height}px`;
      }
    }
  }
  inkBarNodeStyle.display = activeTab ? 'block' : 'none';
}

export default {

  getDefaultProps() {
    return {
      inkBarAnimated: true,
    };
  },

  componentDidUpdate() {
    componentDidUpdate(this);
  },

  componentDidMount() {
    componentDidUpdate(this, true);
  },

  getInkBarNode() {
    const { prefixCls, styles, inkBarAnimated } = this.props;
    const className = `${prefixCls}-ink-bar`;
    const classes = classnames({
      [className]: true,
      [
        inkBarAnimated ?
          `${className}-animated` :
          `${className}-no-animated`
        ]: true,
    });
    return (
      <div
        style={styles.inkBar}
        className={classes}
        key="inkBar"
        ref="inkBar"
      />
    );
  },
};
