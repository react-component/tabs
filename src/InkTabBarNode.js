import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { setTransform, isTransform3dSupported, getLeft, getStyle, getTop, getActiveIndex } from './utils';

function componentDidUpdate(component, init) {
  const { styles, panels, activeKey, direction } = component.props;
  const rootNode = component.props.getRef('root');
  const wrapNode = component.props.getRef('nav') || rootNode;
  const inkBarNode = component.props.getRef('inkBar');
  const activeTab = component.props.getRef('activeTab');
  const inkBarNodeStyle = inkBarNode.style;
  const tabBarPosition = component.props.tabBarPosition;
  const activeIndex = getActiveIndex(panels, activeKey);
  if (init) {
    // prevent mount animation
    inkBarNodeStyle.display = 'none';
  }
  if (activeTab) {
    const tabNode = activeTab;
    const transformSupported = isTransform3dSupported(inkBarNodeStyle);

    // Reset current style
    setTransform(inkBarNodeStyle, '');
    inkBarNodeStyle.width = '';
    inkBarNodeStyle.height = '';
    inkBarNodeStyle.left = '';
    inkBarNodeStyle.top = '';
    inkBarNodeStyle.bottom = '';
    inkBarNodeStyle.right = '';

    if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
      let left = getLeft(tabNode, wrapNode);
      let width = tabNode.offsetWidth;

      // If tabNode'width width equal to wrapNode'width when tabBarPosition is top or bottom
      // It means no css working, then ink bar should not have width until css is loaded
      // Fix https://github.com/ant-design/ant-design/issues/7564
      if (width === rootNode.offsetWidth) {
        width = 0;
      } else if (styles.inkBar && styles.inkBar.width !== undefined) {
        width = parseFloat(styles.inkBar.width, 10);
        if (width) {
          left += (tabNode.offsetWidth - width) / 2;
        }
      }
      if (direction === 'rtl') {
        left = getStyle(tabNode, 'margin-left') - left;
      }
      // use 3d gpu to optimize render
      if (transformSupported) {
        setTransform(inkBarNodeStyle, `translate3d(${left}px,0,0)`);
      } else {
        inkBarNodeStyle.left = `${left}px`;
      }
      inkBarNodeStyle.width = `${width}px`;
    } else {
      let top = getTop(tabNode, wrapNode, true);
      let height = tabNode.offsetHeight;
      if (styles.inkBar && styles.inkBar.height !== undefined) {
        height = parseFloat(styles.inkBar.height, 10);
        if (height) {
          top += (tabNode.offsetHeight - height) / 2;
        }
      }
      if (transformSupported) {
        setTransform(inkBarNodeStyle, `translate3d(0,${top}px,0)`);
        inkBarNodeStyle.top = '0';
      } else {
        inkBarNodeStyle.top = `${top}px`;
      }
      inkBarNodeStyle.height = `${height}px`;
    }
  }
  inkBarNodeStyle.display = activeIndex !== -1 ? 'block' : 'none';
}

export default class InkTabBarNode extends React.Component {
  componentDidMount() {
    // ref https://github.com/ant-design/ant-design/issues/8678
    // ref https://github.com/react-component/tabs/issues/135
    // InkTabBarNode need parent/root ref for calculating position
    // since parent componentDidMount triggered after child componentDidMount
    // we're doing a quick fix here to use setTimeout to calculate position
    // after parent/root component mounted
    this.timeout = setTimeout(() => {
      componentDidUpdate(this, true);
    }, 0);
  }

  componentDidUpdate() {
    componentDidUpdate(this);
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);
  }

  render() {
    const { prefixCls, styles, inkBarAnimated } = this.props;
    const className = `${prefixCls}-ink-bar`;
    const classes = classnames({
      [className]: true,
      [
        inkBarAnimated ?
          `${className}-animated` :
          `${className}-no-animated`
      ]: true,
    });
    return (
      <div
        style={styles.inkBar}
        className={classes}
        key="inkBar"
        ref={this.props.saveRef('inkBar')}
      />
    );
  }
}

InkTabBarNode.propTypes = {
  prefixCls: PropTypes.string,
  styles: PropTypes.object,
  inkBarAnimated: PropTypes.bool,
  saveRef: PropTypes.func,
  direction: PropTypes.string,
};

InkTabBarNode.defaultProps = {
  prefixCls: '',
  inkBarAnimated: true,
  styles: {},
  saveRef: () => { },
};
