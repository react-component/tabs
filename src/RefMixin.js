export default {
  saveRef(name) {
    return node => {
      this[name] = node;
    };
  },
};
