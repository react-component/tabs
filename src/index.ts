import Tabs, { TabsProps } from './Tabs';
import TabPane, { TabPaneProps } from './TabPanelList/TabPane';

export { TabPane, TabsProps, TabPaneProps };

export default Tabs;
