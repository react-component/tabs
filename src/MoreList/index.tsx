import * as React from 'react';
import { useState, useEffect } from 'react';
import KeyCode from 'rc-util/lib/KeyCode';
import Menu, { MenuItem } from 'rc-menu';
import Dropdown from 'rc-dropdown';
import { Tab } from '../interface';

export interface MoreListProps {
  prefixCls: string;
  id: string;
  tabs: Tab[];
  activeKey: string;
  moreIcon?: React.ReactNode;
  onTabClick: (key: React.Key, e: React.MouseEvent | React.KeyboardEvent) => void;
}

export default function MoreList({
  prefixCls,
  id,
  tabs,
  moreIcon = 'More',
  onTabClick,
}: MoreListProps) {
  const [open, setOpen] = useState(false);
  const [selectedKey, setSelectedKey] = useState<string>(null);

  const popupId = `${id}-more-popup`;
  const dropdownPrefix = `${prefixCls}-dropdown`;
  const selectedItemId = selectedKey !== null ? `${popupId}-${selectedKey}` : null;

  const menu = (
    <Menu
      onClick={({ key, domEvent }) => {
        onTabClick(key, domEvent);
        setOpen(false);
      }}
      id={popupId}
      tabIndex={-1}
      role="listbox"
      aria-activedescendant={selectedItemId}
      selectedKeys={[selectedKey]}
    >
      {tabs.map(tab => (
        <MenuItem
          key={tab.key}
          id={`${popupId}-${tab.key}`}
          role="option"
          aria-controls={id && `${id}-panel-${tab.key}`}
        >
          {tab.tab}
        </MenuItem>
      ))}
    </Menu>
  );

  function selectOffset(offset: -1 | 1) {
    const enabledTabs = tabs.filter(tab => !tab.disabled);
    let selectedIndex = enabledTabs.findIndex(tab => tab.key === selectedKey) || 0;
    const len = enabledTabs.length;

    if (len > 1) {
      selectedIndex = (selectedIndex + offset + len) % len;
      setSelectedKey(enabledTabs[selectedIndex].key);
    }
  }

  function onKeyDown(e: React.KeyboardEvent) {
    const { which } = e;

    if (!open) {
      if (which === KeyCode.DOWN) {
        setOpen(true);
      }
      return;
    }

    switch (which) {
      case KeyCode.UP:
        selectOffset(-1);
        break;
      case KeyCode.DOWN:
        selectOffset(1);
        break;
      case KeyCode.ESC:
        setOpen(false);
        break;
      case KeyCode.SPACE:
      case KeyCode.ENTER:
        if (selectedKey !== null) onTabClick(selectedKey, e);
        break;

      default:
      // Do nothing
    }
  }

  // We use query element here to avoid React strict warning
  useEffect(() => {
    const ele = document.getElementById(selectedItemId);
    if (ele) {
      ele.scrollIntoView(false);
    }
  }, [selectedKey]);

  useEffect(() => {
    if (!open) {
      setSelectedKey(null);
    }
  }, [open]);

  return (
    <Dropdown
      prefixCls={dropdownPrefix}
      overlay={menu}
      trigger={['click']}
      visible={open}
      onVisibleChange={setOpen}
    >
      <button
        type="button"
        className={`${prefixCls}-nav-more`}
        style={{ visibility: tabs.length ? null : 'hidden' }}
        aria-haspopup="listbox"
        aria-controls={popupId}
        id={`${id}-more`}
        aria-expanded={open}
        onKeyDown={onKeyDown}
      >
        {moreIcon}
      </button>
    </Dropdown>
  );
}
