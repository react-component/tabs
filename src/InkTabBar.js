import createReactClass from 'create-react-class';
import InkTabBarMixin from './InkTabBarMixin';
import TabBarMixin from './TabBarMixin';

const InkTabBar = createReactClass({
  displayName: 'InkTabBar',
  activeTab: null,
  mixins: [TabBarMixin, InkTabBarMixin],
  render() {
    const inkBarNode = this.getInkBarNode();
    const tabs = this.getTabs();
    return this.getRootNode([inkBarNode, tabs]);
  },
});

export default InkTabBar;
