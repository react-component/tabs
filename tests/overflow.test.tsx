import React from 'react';
import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import { act } from 'react-dom/test-utils';
import {
  getOffsetSizeFunc,
  getTabs,
  triggerResize,
  getTransformX,
  getTransformY,
} from './common/util';
import { TabPane } from '../src';

describe('Tabs.Overflow', () => {
  let domSpy: ReturnType<typeof spyElementPrototypes>;
  let holder: HTMLDivElement;

  const hackOffsetInfo: { list?: number } = {};

  beforeAll(() => {
    holder = document.createElement('div');
    document.body.appendChild(holder);

    function btnOffsetPosition() {
      const btn = this as HTMLButtonElement;
      const btnList = [...btn.parentNode.childNodes].filter(ele =>
        (ele as HTMLElement).className.includes('rc-tabs-tab'),
      );
      const index = btnList.indexOf(btn);
      return 20 * index;
    }

    domSpy = spyElementPrototypes(HTMLElement, {
      scrollIntoView: () => {},
      offsetWidth: {
        get: getOffsetSizeFunc(hackOffsetInfo),
      },
      offsetHeight: {
        get: getOffsetSizeFunc(hackOffsetInfo),
      },
      offsetLeft: {
        get: btnOffsetPosition,
      },
      offsetTop: {
        get: btnOffsetPosition,
      },
    });
  });

  afterAll(() => {
    domSpy.mockRestore();
    document.body.removeChild(holder);
  });

  it('should collapse', () => {
    jest.useFakeTimers();
    const onChange = jest.fn();
    const wrapper = mount(getTabs({ onChange }));

    triggerResize(wrapper);
    act(() => {
      jest.runAllTimers();
      wrapper.update();
    });
    expect(wrapper.find('.rc-tabs-nav-more').render()).toMatchSnapshot();

    // Click to open
    wrapper.find('.rc-tabs-nav-more').simulate('mouseenter');
    jest.runAllTimers();
    wrapper.update();
    expect(
      wrapper
        .find('.rc-tabs-dropdown li')
        .first()
        .text(),
    ).toEqual('cute');

    // Click to select
    wrapper
      .find('.rc-tabs-dropdown-menu-item')
      .first()
      .simulate('click');
    expect(onChange).toHaveBeenCalledWith('cute');

    wrapper.unmount();

    jest.useRealTimers();
  });

  [KeyCode.SPACE, KeyCode.ENTER].forEach(code => {
    it(`keyboard with select keycode: ${code}`, () => {
      jest.useFakeTimers();
      const onChange = jest.fn();
      const wrapper = mount(getTabs({ onChange }), { attachTo: holder });

      triggerResize(wrapper);
      act(() => {
        jest.runAllTimers();
        wrapper.update();
      });

      // Open
      wrapper.find('.rc-tabs-nav-more').simulate('keydown', {
        which: KeyCode.DOWN,
      });

      // key selection
      function keyMatch(which: number, match: string) {
        wrapper.find('.rc-tabs-nav-more').simulate('keydown', {
          which,
        });
        expect(wrapper.find('.rc-tabs-dropdown-menu-item-selected').text()).toEqual(match);
      }

      keyMatch(KeyCode.DOWN, 'cute');
      keyMatch(KeyCode.DOWN, 'miu');
      keyMatch(KeyCode.UP, 'cute');

      // Select
      wrapper.find('.rc-tabs-nav-more').simulate('keydown', {
        which: code,
      });
      expect(onChange).toHaveBeenCalledWith('cute');

      // Open
      wrapper.find('.rc-tabs-nav-more').simulate('keydown', {
        which: KeyCode.DOWN,
      });
      wrapper.update();
      expect(
        wrapper
          .find('.rc-tabs-dropdown')
          .last()
          .hasClass('rc-tabs-dropdown-hidden'),
      ).toBeFalsy();

      // ESC
      wrapper.find('.rc-tabs-nav-more').simulate('keydown', {
        which: KeyCode.ESC,
      });
      wrapper.update();
      expect(
        wrapper
          .find('.rc-tabs-dropdown')
          .last()
          .hasClass('rc-tabs-dropdown-hidden'),
      ).toBeTruthy();

      wrapper.unmount();

      jest.useRealTimers();
    });
  });

  describe('wheel', () => {
    const list: { name: string; x1: number; y1: number; x2: number; y2: number }[] = [
      {
        name: 'deltaX',
        x1: 20,
        y1: 5,
        x2: 3,
        y2: -3,
      },
      {
        name: 'deltaY',
        y1: 20,
        x1: 5,
        y2: 3,
        x2: -3,
      },
    ];

    ['top', 'left'].forEach((tabPosition: any) => {
      list.forEach(({ name, x1, y1, x2, y2 }) => {
        it(`should ${tabPosition} work for ${name}`, () => {
          jest.useFakeTimers();
          const wrapper = mount(getTabs({ tabPosition }), { attachTo: holder });

          triggerResize(wrapper);
          act(() => {
            jest.runAllTimers();
            wrapper.update();
          });

          // Wheel to move
          const node = (wrapper.find('.rc-tabs-nav-wrap').instance() as unknown) as HTMLElement;

          act(() => {
            const touchStart = new WheelEvent('wheel', {
              deltaX: x1,
              deltaY: y1,
            });
            node.dispatchEvent(touchStart);
            jest.runAllTimers();
          });

          act(() => {
            const touchStart = new WheelEvent('wheel', {
              deltaX: x2,
              deltaY: y2,
            });
            node.dispatchEvent(touchStart);
            jest.runAllTimers();
          });

          wrapper.update();
          if (tabPosition === 'top') {
            expect(getTransformX(wrapper)).toEqual(-23);
          } else {
            expect(getTransformY(wrapper)).toEqual(-23);
          }

          wrapper.unmount();
          jest.useRealTimers();
        });
      });
    });

    ['top', 'left'].forEach((tabPosition: any) => {
      it(`no need if place is enough: ${tabPosition}`, () => {
        hackOffsetInfo.list = 20;

        jest.useFakeTimers();
        const wrapper = mount(getTabs({ children: [<TabPane key="yo">Yo</TabPane>], tabPosition }));

        triggerResize(wrapper);
        act(() => {
          jest.runAllTimers();
          wrapper.update();
        });

        // Wheel to move
        const node = (wrapper.find('.rc-tabs-nav-wrap').instance() as unknown) as HTMLElement;
        const touchStart = new WheelEvent('wheel', {
          deltaX: 20,
          deltaY: 20,
        });
        touchStart.preventDefault = jest.fn();

        act(() => {
          node.dispatchEvent(touchStart);
          jest.runAllTimers();
        });

        expect(touchStart.preventDefault).not.toHaveBeenCalled();
        expect(getTransformX(wrapper)).toEqual(0);

        jest.useRealTimers();
        hackOffsetInfo.list = undefined;
      });
    });
  });

  describe('overflow to scroll', () => {
    it('top', () => {
      jest.useFakeTimers();
      const onTabScroll = jest.fn();
      // light bamboo [cute disabled] miu
      const wrapper = mount(getTabs({ activeKey: 'disabled', onTabScroll }));

      triggerResize(wrapper);
      act(() => {
        jest.runAllTimers();
        wrapper.update();
      });
      expect(getTransformX(wrapper)).toEqual(-40);

      // light [bamboo cute] disabled miu
      onTabScroll.mockReset();
      wrapper.setProps({ activeKey: 'bamboo' });
      act(() => {
        jest.runAllTimers();
        wrapper.update();
      });
      expect(getTransformX(wrapper)).toEqual(-20);
      expect(onTabScroll).toHaveBeenCalledWith({ direction: 'left' });

      jest.useRealTimers();
    });

    it('left', () => {
      jest.useFakeTimers();
      const onTabScroll = jest.fn();
      /**
       *    light        light
       *    bamboo      --------
       *   --------      bamboo
       *     cute         cute
       *   disabled     --------
       *   --------     disabled
       *     miu          miu
       */
      const wrapper = mount(getTabs({ activeKey: 'disabled', tabPosition: 'left', onTabScroll }));

      triggerResize(wrapper);
      act(() => {
        jest.runAllTimers();
        wrapper.update();
      });
      expect(getTransformY(wrapper)).toEqual(-40);

      // light [bamboo cute] disabled miu
      onTabScroll.mockReset();
      wrapper.setProps({ activeKey: 'bamboo' });
      act(() => {
        jest.runAllTimers();
        wrapper.update();
      });
      expect(getTransformY(wrapper)).toEqual(-20);
      expect(onTabScroll).toHaveBeenCalledWith({ direction: 'top' });

      jest.useRealTimers();
    });
  });
});
