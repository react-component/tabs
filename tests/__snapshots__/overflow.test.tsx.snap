// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.Overflow should collapse 1`] = `
<button
  aria-controls="rc-tabs-test-more-popup"
  aria-expanded="false"
  aria-haspopup="listbox"
  aria-hidden="true"
  class="rc-tabs-nav-more"
  id="rc-tabs-test-more"
  style=""
  tabindex="-1"
  type="button"
>
  More
</button>
`;
